service/history/timer_queue_active_task_executor.go

```diff
@@ -711,6 +711,52 @@ func (t *timerQueueActiveTaskExecutor) executeWorkflowExecutionTimeoutTask(
 	return nil
 }
 
+func (t *timerQueueActiveTaskExecutor) executeStateMachineTimerTask(
+	ctx context.Context,
+	task *tasks.StateMachineTimerTask,
+	execute func(node *hsm.Node, task hsm.Task) error,
+) (retError error) {
+	ctx, cancel := context.WithTimeout(ctx, taskTimeout)
+	defer cancel()
+
+	wfCtx, release, err := getWorkflowExecutionContextForTask(ctx, t.shardContext, t.cache, task)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	ms, err := loadMutableStateForTimerTask(ctx, t.shardContext, wfCtx, task, t.metricsHandler, t.logger)
+	if err != nil {
+		return err
+	}
+	if ms == nil {
+		return nil
+	}
+
+	timers, err := t.executeStateMachineTimers(ctx, ms, execute)
+	if err != nil {
+		return err
+	}
+
+	// We haven't done any work, return without committing.
+	if len(timers) == len(ms.GetExecutionInfo().StateMachineTimers) {
+		return nil
+	}
+
+	// Update processed timers and ensure the next timer task is scheduled.
+	ms.GetExecutionInfo().StateMachineTimers = timers
+	// TODO(bergundy): Right now there's an early return in task_generator.go GenerateDirtySubStateMachineTasks because
+	// transition history may be disabled. Remove this line from here once that early return is gone.
+	workflow.AddNextStateMachineTimerTask(ms)
```

Comment:
I'd move this into the shared executeStateMachineTimers code but have that method return how many timers were executed.